src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,197 @@
+                    }
+
+                    if (Interlocked.CompareExchange(ref referenceCount.Value, currentValue + 1, currentValue) == currentValue)
+                    {
+                        return new ReferenceCountedDisposable<T>(target, referenceCount);
+                    }
+                }
+            }
+
+            public void Dispose()
+            {
+                var referenceCount = Interlocked.Exchange(ref _referenceCount, null);
+                if (referenceCount == null)
+                {
+                    // Already disposed; allow multiple without error.
+                    return;
+                }
+
+                // This is the thread which will write to _instance, so this will act as an atomic read
+                var instance = _instance;
+
+                // Set the instance back to its default value, so in case someone forgets to dispose of one of the
+                // counted references the finalizer of the underlying disposable object will have a chance to clean up
+                // the unmanaged resources as soon as possible.
+                Interlocked.MemoryBarrier();
+                _instance = default(T);
+
+#if !DEBUG
+                var decrementedValue = Interlocked.Decrement(ref referenceCount.Value);
+                if (decrementedValue == 0)
+                {
+                    instance.Dispose();
+                }
+#else
+                while (true)
+                {
+                    var currentValue = Volatile.Read(ref referenceCount.Value);
+                    Debug.Assert(currentValue > 0, "Dispose should have protected itself against races.");
+
+                    if (Interlocked.CompareExchange(ref referenceCount.Value, currentValue - 1, currentValue) == currentValue)
+                    {
+                        if (currentValue == 1)
+                        {
+                            // Reference count hit 0 for this call
+                            instance.Dispose();
+                        }
+
+                        break;
+                    }
+                }
+#endif
+            }
+
+            /// <summary>
+            /// Represents a weak reference to a <see cref="ReferenceCountedDisposable{T}"/> which is capable of
+            /// obtaining a new counted reference up until the point when the object is no longer accessible.
+            /// </summary>
+            public struct WeakReference
+            {
+                /// <summary>
+                /// DO NOT DISPOSE OF THE TARGET.
+                /// </summary>
+                private readonly WeakReference<ReferenceCountedDisposable<T>> _instance;
+
+                public WeakReference(ReferenceCountedDisposable<T> reference)
+                    : this()
+                {
+                    var instance = reference._instance;
+                    Interlocked.MemoryBarrier();
+                    var referenceCount = Volatile.Read(ref reference._referenceCount);
```

Comment:
having _referenceCount not be an int makes this all very strange.  Can you call it '_referenceCountStrongBox' or something to make it clearer?  Otherwise seeing comparisons to null is quite odd.